src/app/modules/main/wallet_section/all_tokens/flat_tokens_model.nim

```diff
@@ -0,0 +1,116 @@
+    # built from chainId and address using networks service
+    WebsiteUrl
+    MarketValues
+
+QtObject:
+  type FlatTokensModel* = ref object of QAbstractListModel
+    delegate: io_interface.FlatTokenModelDataSource
+
+  proc setup(self: FlatTokensModel) =
+    self.QAbstractListModel.setup
+
+  proc delete(self: FlatTokensModel) =
+    self.QAbstractListModel.delete
+
+  proc newFlatTokensModel*(delegate: io_interface.FlatTokenModelDataSource): FlatTokensModel =
+    new(result, delete)
+    result.setup
+    result.delegate = delegate
+
+  method rowCount(self: FlatTokensModel, index: QModelIndex = nil): int =
+    return self.delegate.getFlatTokensList().len
+
+  proc countChanged(self: FlatTokensModel) {.signal.}
+  proc getCount(self: FlatTokensModel): int {.slot.} =
+    return self.rowCount()
+  QtProperty[int] count:
+    read = getCount
+    notify = countChanged
+
+  method roleNames(self: FlatTokensModel): Table[int, string] =
+    {
+      ModelRole.Key.int:"key",
+      ModelRole.Name.int:"name",
+      ModelRole.Symbol.int:"symbol",
+      ModelRole.Sources.int:"sources",
+      ModelRole.ChainId.int:"chainId",
+      ModelRole.Address.int:"address",
+      ModelRole.Decimals.int:"decimals",
+      ModelRole.Image.int:"image",
+      ModelRole.Type.int:"type",
+      ModelRole.CommunityId.int:"communityId",
+      ModelRole.Description.int:"description",
+      ModelRole.WebsiteUrl.int:"websiteUrl",
+      ModelRole.MarketValues.int:"marketValues",
+    }.toTable
+
+  method data(self: FlatTokensModel, index: QModelIndex, role: int): QVariant =
+    if not index.isValid:
+      return
+    if index.row < 0 or index.row >= self.rowCount():
+      return
+    # the only way to read items from service is by this single method getFlatTokensList
+    let item = self.delegate.getFlatTokensList()[index.row]
+    let enumRole = role.ModelRole
+    case enumRole:
+      of ModelRole.Key:
+        result = newQVariant(item.key)
+      of ModelRole.Name:
+        result = newQVariant(item.name)
+      of ModelRole.Symbol:
+        result = newQVariant(item.symbol)
+      of ModelRole.Sources:
+        result = newQVariant(SOURCES_DELIMITER & item.sources.join(SOURCES_DELIMITER) & SOURCES_DELIMITER)
+      of ModelRole.ChainId:
+        result = newQVariant(item.chainId)
+      of ModelRole.Address:
+        result = newQVariant(item.address)
+      of ModelRole.Decimals:
+        result = newQVariant(item.decimals)
+      of ModelRole.Image:
+        result = newQVariant(item.image)
+      of ModelRole.Type:
+        result = newQVariant(ord(item.`type`))
+      of ModelRole.CommunityId:
+        result = newQVariant(item.communityId)
+      # ToDo fetching of market values not done yet
+      of ModelRole.Description:
+        result = newQVariant("")
+      of ModelRole.WebsiteUrl:
+        result = newQVariant("")
+      of ModelRole.MarketValues:
+        result = newQVariant("")
```

Comment:
Uhm, why are these always empty?

Reply:
hey this is a todo -> https://github.com/status-im/status-desktop/issues/12668 will covered here